networkx/relabel.py

```diff
@@ -116,44 +118,38 @@ def relabel_nodes(G, mapping, copy=True):
     if not hasattr(mapping, "__getitem__"):
         m = {n: mapping(n) for n in G}
     else:
-        m = mapping
+        # enforce a complete mapping (if only partial mapping is defined)
+        m = {n: mapping.get(n, n) for n in G}
     if copy:
         return _relabel_copy(G, m)
     else:
         return _relabel_inplace(G, m)
 
 
 def _relabel_inplace(G, mapping):
-    old_labels = set(mapping.keys())
-    new_labels = set(mapping.values())
-    if len(old_labels & new_labels) > 0:
+    if len(mapping.keys() & mapping.values()) > 0:
         # labels sets overlap
         # can we topological sort and still do the relabeling?
         D = nx.DiGraph(list(mapping.items()))
         D.remove_edges_from(nx.selfloop_edges(D))
         try:
-            nodes = reversed(list(nx.topological_sort(D)))
+            # only keep the nodes in the original graph
+            nodes = [n for n in nx.topological_sort(D) if n in mapping.keys()]
```

Comment:
This change should not be made. The topological sort is needed to handle the order of the changes that allow in-place relabeling to work.  e.g. if we change `1 -> 3` and then later change `3->4`, that's not what the mapping asked for.  I think we cannot leave the node order unchanged if we have an in-place relabeling with overlapping node sets.